elastic/kibana
# [Stateful sidenav] Activation mechanism

In this PR I've introduced **part** of the activation mechanism for the side navigation evolution. There will be a following work on the activation to save the state in the user profile.
As a reminder, the side navigation evolution is the work to bring the serverless UX navigation to stateful Kibana.

The main purpose of this PR is to introduce the global feature flag to enable the new side navigation on a deployment. Future UI work will then be possible to be merged into `main` behind that feature flag.

## Notes on implementation

**Important:** The conversation is still ongoing on the exact activation mechanism regarding whether admins will have all the settings proposed in this PR. There will be future polish works around text copy when the final decision is made around whic advanced settings are exposed.

### New `kibana.yml` settings

I've introduced the following settings in `kibana.yml`:

```yaml
# Feature flag for the new side navigation. If it is not set or set to false,
# the new side navigation will not be visible anywhere in the Kibana UI.
navigation.solutionNavigation.featureOn: true|false

# Once the feature flag is enabled, this settings controls if the new side navigation
# is enabled of not. It maps to a Kibana advanced setting.
navigation.solutionNavigation.enabled: true

# If the above 'enabled' setting is set to true, this setting defines how user
# opt-in experience is handle.
# - visible: the user immediately sees the new side navigation
# - hidden: the user does not see the new side navigation. They can opt-in from their user profile.
# - ask: the user is asked if they want to opt-in to the new side navigation.
navigation.solutionNavigation.optInStatus: visible|hidden|ask

# This setting defines the default solution that the user sees when the new side navigation is enabled.
# If set to "ask", the user will be asked to choose a solution once they opt-in.
navigation.solutionNavigation.defaultSolution: es|oblt|security|ask
```

### Advanced settings

Once the `navigation.solutionNavigation.featureOn` is set to `true`, the following global advanced settings will be available to administrators. As you can see, the settings are the same as the ones in `kibana.yml`:

<img width="950" alt="Screenshot 2024-02-23 at 13 38 04" src="https://github.com/elastic/kibana/assets/2854616/d85db2bb-249b-4a8f-9b78-0c497c8421d8">


## How to test


* Launch kibana and navigate to advanced settings (`app/management/kibana/settings`)
* Select the "Global settings" tab and make sure that there is no "Technical preview" section
* Add the following settings to your `kibana.yml`:

```yaml
navigation.solutionNavigation.featureOn: true
```

* Refresh the browser window
* In the "Global advanced settings" tab, you should see the "Technical preview" section with the default settings from the `kibana.yml` file (`enabled`, `optInStatus`, `defaultSolution`)
* Change the yaml setting to `false`, refresh the browser window and check that the "Technical preview" section is gone
* Change the yaml setting back to `true` and refresh the browser window
* Change the "Enable solution navigation" setting to `true` and the "Opt-in behaviour" to `visible`.
* Save the changes
* You should immediately see the new side navigation

## Screenshot

<img width="1466" alt="Screenshot 2024-02-23 at 13 37 41" src="https://github.com/elastic/kibana/assets/2854616/b0f8db5a-5f08-4bcb-83b4-fecb9f4d1479">

Fixes https://github.com/elastic/kibana/issues/176810
